### PR TITLE
KTOR-3201 Migrate plugins to subroute

### PR DESCRIPTION
**Pull request by rsinukov**
And clean up constructors

### CODE REVIEW

**Comment by e5l** [comment]
@andreyaksenov, could you check the docs?

**Review by e5l — APPROVED** [review]
Sorry for the delay, LGTM

File: ktor-server/ktor-server-plugins/ktor-server-caching-headers/jvm/src/io/ktor/server/plugins/CachingHeaders.kt [hunk 1]
```diff
@@ -15,7 +15,9 @@ import io.ktor.util.pipeline.*
  * Plugin that set [CachingOptions] headers for every response.
  * It invokes [optionsProviders] for every response and use first non null caching options
  */
-public class CachingHeaders(private val optionsProviders: List<(OutgoingContent) -> CachingOptions?>) {
+public class CachingHeaders private constructor(
+    private val optionsProviders: List<(OutgoingContent) -> CachingOptions?>
+) {
     /**
      * Configuration for [CachingHeaders] plugin
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-caching-headers/jvm/src/io/ktor/server/plugins/CachingHeaders.kt:22]: > A configuration for the [CachingHeaders] plugin

File: ktor-server/ktor-server-plugins/ktor-server-call-id/jvm/src/io/ktor/server/plugins/CallId.kt [hunk 2]
```diff
@@ -180,7 +180,7 @@ public class CallId private constructor(
     /**
      * Installable plugin for [CallId]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-call-id/jvm/src/io/ktor/server/plugins/CallId.kt:181]: > An installable plugin for [CallId]

File: ktor-server/ktor-server-plugins/ktor-server-call-id/jvm/src/io/ktor/server/plugins/CallId.kt [hunk 3]
```diff
@@ -180,7 +180,7 @@ public class CallId private constructor(
...
      */
-    public companion object Plugin : ApplicationPlugin<ApplicationCallPipeline, Configuration, CallId> {
+    public companion object Plugin : RouteScopedPlugin<Configuration, CallId> {
         /**
          * [ApplicationCallPipeline]'s phase which this plugin will be installed to
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-call-id/jvm/src/io/ktor/server/plugins/CallId.kt:185]: > An [ApplicationCallPipeline]'s phase to which this plugin is installed

File: ktor-server/ktor-server-plugins/ktor-server-compression/jvm/src/io/ktor/server/plugins/Compression.kt [hunk 4]
```diff
@@ -61,7 +61,7 @@ public data class CompressionEncoderConfig(
 /**
  * Plugin to compress a response based on conditions and ability of client to decompress it
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-compression/jvm/src/io/ktor/server/plugins/Compression.kt:62]: > A plugin for compressing a response based on specific conditions and the ability of the client to decompress it

File: ktor-server/ktor-server-plugins/ktor-server-conditional-headers/jvm/src/io/ktor/server/plugins/ConditionalHeaders.kt [hunk 5]
```diff
@@ -16,7 +16,9 @@ import io.ktor.util.pipeline.*
 /**
  * Plugin to check modified/match conditional headers and avoid sending contents if it was not changed
  */
-public class ConditionalHeaders(private val versionProviders: List<suspend (OutgoingContent) -> List<Version>>) {
+public class ConditionalHeaders private constructor(
+    private val versionProviders: List<suspend (OutgoingContent) -> List<Version>>
+) {
 
     /**
      * Configuration for [ConditionalHeaders] plugin
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-conditional-headers/jvm/src/io/ktor/server/plugins/ConditionalHeaders.kt:24]: > A configuration for the [ConditionalHeaders] plugin

File: ktor-server/ktor-server-plugins/ktor-server-freemarker/jvm/src/io/ktor/server/freemarker/FreeMarker.kt [hunk 6]
```diff
@@ -30,7 +30,7 @@ public class FreeMarkerContent(
 /**
  * Freemarker support plugin. Provides ability to respond with [FreeMarkerContent]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-freemarker/jvm/src/io/ktor/server/freemarker/FreeMarker.kt:31]: > A plugin that allows you to use FreeMarker templates as views within your application. Provides the ability to respond with [FreeMarkerContent]

File: ktor-server/ktor-server-plugins/ktor-server-hsts/jvm/src/io/ktor/server/plugins/HSTS.kt [hunk 7]
```diff
@@ -81,7 +81,7 @@ public class HSTS(config: Configuration) {
     /**
      * Plugin installation object
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-hsts/jvm/src/io/ktor/server/plugins/HSTS.kt:82]: > A plugin installation object

File: ktor-server/ktor-server-plugins/ktor-server-http-redirect/jvm/src/io/ktor/server/plugins/HttpsRedirect.kt [hunk 8]
```diff
@@ -13,7 +13,7 @@ import io.ktor.util.*
 /**
  * Redirect non-secure requests to HTTPS
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-http-redirect/jvm/src/io/ktor/server/plugins/HttpsRedirect.kt:14]: > A plugin that redirects non-secure HTTP calls to HTTPS

File: ktor-server/ktor-server-plugins/ktor-server-pebble/jvm/src/io/ktor/server/pebble/Pebble.kt [hunk 9]
```diff
@@ -36,7 +36,7 @@ public class PebbleContent(
 /**
  * Plugin for providing Pebble templates as [PebbleContent]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-pebble/jvm/src/io/ktor/server/pebble/Pebble.kt:37]: > A plugin that allows you to use Pebble templates as views within your application. Provides the ability to respond with [PebbleContent]

File: ktor-server/ktor-server-plugins/ktor-server-sessions/jvm/src/io/ktor/server/sessions/Sessions.kt [hunk 10]
```diff
@@ -13,7 +13,7 @@ import kotlin.reflect.*
  * Sessions plugin that provides a mechanism to persist information between requests.
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-sessions/jvm/src/io/ktor/server/sessions/Sessions.kt:13]: > A plugin that provides a mechanism to persist data between different HTTP requests.

File: ktor-server/ktor-server-plugins/ktor-server-status-pages/jvm/src/io/ktor/server/plugins/StatusPages.kt [hunk 11]
```diff
@@ -16,7 +16,7 @@ import kotlinx.coroutines.*
 /**
  * Status pages plugin that handles exceptions and status codes. Useful to configure default error pages.
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-status-pages/jvm/src/io/ktor/server/plugins/StatusPages.kt:17]: > A plugin that handles ...

File: ktor-server/ktor-server-plugins/ktor-server-thymeleaf/jvm/src/io/ktor/server/thymeleaf/Thymeleaf.kt [hunk 12]
```diff
@@ -33,7 +33,7 @@ public class ThymeleafContent(
 /**
  * Thymeleaf support plugin. Provides ability to respond with [Thymeleaf]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-thymeleaf/jvm/src/io/ktor/server/thymeleaf/Thymeleaf.kt:34]: > A plugin that allows you to use Thymeleaf templates as views within your application. Provides the ability to respond with [ThymeleafContent]

File: ktor-server/ktor-server-plugins/ktor-server-velocity/jvm/src/io/ktor/server/velocity/Velocity.kt [hunk 13]
```diff
@@ -48,7 +48,7 @@ internal fun velocityOutgoingContent(
 /**
  * Velocity ktor plugin. Provides ability to respond with [VelocityContent] and [respondTemplate].
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-plugins/ktor-server-velocity/jvm/src/io/ktor/server/velocity/Velocity.kt:49]: > A plugin that allows you to use Velocity templates as views within your application. Provides the ability to respond with [VelocityContent].

**Review by andreyaksenov — CHANGES_REQUESTED** [review]

**Review by andreyaksenov — APPROVED** [review]

**Pull request merged**